### PR TITLE
[service] Introduce network server and client

### PR DESCRIPTION
**Pull request by JingsongLi**
### Purpose

Introduce a network framework for service. (can be used by query service and log service)

Most of parts are refer to Flink Queryable State:
https://github.com/apache/flink/tree/master/flink-queryable-state

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/lookup/QueryServer.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+package org.apache.paimon.lookup;
+
+import java.net.InetSocketAddress;
+
+/**
+ * An interface for the Server running on each Node in the cluster. This server is responsible for
+ * serving requests coming from the client.
+ */
+public interface QueryServer {
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/lookup/QueryServer.java:27]: The implementation should be in paimon-core, lookup capability should in paimon-core.

paimon-service-runtime should only include client server things, just like RPC.

File: paimon-service/paimon-service-client/pom.xml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            <groupId>org.apache.paimon</groupId>
+            <artifactId>paimon-core</artifactId>
+            <version>${project.version}</version>
+            <scope>provided</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.flink</groupId>
+            <artifactId>flink-shaded-netty</artifactId>
+            <version>4.1.91.Final-${flink.shaded.version}</version>
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply paimon-service/paimon-service-client/pom.xml:56]: In future, yes, but now, we will just rely on Flink's netty.

File: paimon-service/paimon-service-client/src/main/java/org/apache/paimon/service/network/NetworkServer.java
Comment (leaves12138) [inline paimon-service/paimon-service-client/src/main/java/org/apache/paimon/service/network/NetworkServer.java:1]: Does this class put in runtime more suitable?

Comment (JingsongLi) [reply paimon-service/paimon-service-client/src/main/java/org/apache/paimon/service/network/NetworkServer.java:1]: Maybe there is some proxy in future. It needs network server.

**Comment by leaves12138** [comment]
The KvRequest now must specify `Partition` `bucket` field, we could only handle foreign key widen by this structure. If the join operator dimension table does not happen in a primary key, or the on clause does not hit the table primary key, the query service can't work with that. Need we consider there situation now?

**Comment by JingsongLi** [comment]
> The `KvRequest` now must specify `Partition` `bucket` field, we could only handle foreign key widen by this structure. If the join operator dimension table does not happen in a primary key, or the `on` clause does not hit the table primary key, the query service can't work with that. Need we consider there situations now?

Good question, yes, foreign key should be considered too, but I found it is very hard to consider it now. It will make design very hard to develop.

And, we do not rely on foreign key, we can use partial-update to implement streaming join.

**Review by tsreaper — APPROVED** [review]

**Pull request merged**